client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -200,8 +210,144 @@ private Checkpoint completeCheckpoint(String checkpointName) {
 
     @Override
     public void resetReaderGroup(ReaderGroupConfig config) {
+        while (true) {
+            val state = getState();
+            switch (state.getConfigState()) {
+                case INITIALIZING:
+                    doInit(state);
+                    break;
+                case REINITIALIZING:
+                    doReinit(state);
+                    break;
+                case READY:
+                    long newGen = state.getGeneration() + 1;
+                    // If true then at this point the update is completed and it is in the desired state (REINITIALIZING)
+                    // on which doReinit is called to continue the reset with this newState.
+                    if (doStateTransition(state, new ReaderGroupStateResetStart(config, newGen))) {
+                        val newState = getState();
+                        doReinit(newState);
+                        return;
+                    }
+                    break;
+                case DELETING:
+                    doDelete(state);
+                    throw new ReinitializationRequiredException("This ReaderGroup does not exist.");
+            }
+        }
+    }
+
+    public void createState(ReaderGroupConfig config) {
         Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, config);
-        synchronizer.updateStateUnconditionally(new ReaderGroupStateInit(config, segments, getEndSegmentsForStreams(config)));
+        synchronizer.initialize(new ReaderGroupState.ReaderGroupStateInit(config, segments, getEndSegmentsForStreams(config)));
+        val state = getState();
+        if (state.getConfigState() == ReaderGroupState.ConfigState.INITIALIZING && state.getConfig().equals(config)) {
+            doInit(state);
+        }
+    }
+
+    public void deleteState() {
+        while (true) {
+            val state = getState();
+            switch (state.getConfigState()) {
```

Comment:
This switch block is similar to the one in the previous method. I'm wondering if there is a better way to implement this to  make more explicit the transitions of the state machine defined by these config states.